src/adaXT/random_forest/RandomForest.py

```diff
@@ -0,0 +1,355 @@
+        return np.array(tree.predict(self.predict_values.read()))
+
+    # Function to call predict on all the trees of the forrest, differentiates
+    # between running in parallel and sequential
+    def __predict_trees(self):
+        predictions = []
+
+        if (self.n_jobs == 1):
+            for tree in self.trees:
+                predictions.append(self._predict_single_tree(tree))
+        else:
+            with Pool(self.n_jobs) as p:
+                predictions = p.map(self._predict_single_tree, self.trees)
+
+        return np.column_stack(predictions)
+
+    # Function used to add a column with zeros for all the classes that are in
+    # the forrest but not in a given tree
+    def __fill_with_zeros_for_missing_classes_in_tree(
+            self, tree_classes, predict_proba, num_rows_predict):
+        ret_val = np.zeros((num_rows_predict, len(self.classes)))
+
+        # Find the indices of tree_classes in forrest_classes
+        tree_class_indices = np.searchsorted(self.classes, tree_classes)
+
+        # Only update columns corresponding to tree_classes
+        ret_val[:, tree_class_indices] = predict_proba
+
+        return ret_val
+
+    # Function to call predict_proba for a tree
+    def _predict_proba_single_tree(self, tree: DecisionTree):
+        tree_predict_proba = tree.predict_proba(self.predict_values.read())
+        ret_val = self.__fill_with_zeros_for_missing_classes_in_tree(
+            tree.classes, tree_predict_proba, self.predict_values._shape[0])
+        return ret_val
+
+    # Function to call predict_proba on all the trees of the forrest,
+    # differentiates between running in parallel and sequential
+    def __predict_proba_trees(self):
+        predictions = []
+
+        if (self.n_jobs == 1):
+            for tree in self.trees:
+                predictions.append(self._predict_proba_single_tree(tree))
+        else:
+            with Pool(self.n_jobs) as p:
+                predictions = p.map(
+                    self._predict_proba_single_tree, self.trees)
+
+        return predictions
+
+    def fit(self, X: np.ndarray, Y: np.ndarray):
+        """
+        Function used to fit a forrest using many DecisionTrees for the given data
+
+        Parameters
+        ----------
+        X : np.ndarray
+            feature values
+        Y : np.ndarray
+            response values
+        """
+        if not self.bootstrap and self.max_samples:
+            raise AttributeError(
+                "Bootstrap can not be False while max_samples is set")
+
+        X, Y = self.check_input(X, Y)
+        self.features = SharedNumpyArray(X)
+        self.outcomes = SharedNumpyArray(Y)
+        self.n_obs, self.n_features = self.features._shape
+        if self.forrest_type == "Classification":
+            self.classes = np.unique(self.outcomes.read())
+        if self.max_samples is None:
+            self.max_samples = self.n_obs
+        self.trees = [
```

Comment:
Might be an idea to do this with the Pool aswell. Seeing that
some people might initialize more than 100 trees,
this could lead to quite a slow setup time without.

Reply:
I think the setup of Pool would require way more time than the initialization of the trees. I think any feasible amount of trees can be initialized sequentially in a minuscule amount of time.